mountpoint-s3-fs/src/superblock.rs

```diff
@@ -251,59 +251,48 @@ impl Superblock {
-                if let Some(parent) = self.inner.inodes.read().unwrap().get(&inode.parent()).cloned() {
-                    parent
+        let mut inodes = self.inner.inodes.write().unwrap();
+
+        match inodes.decrease_lookup_count(ino, n) {
+            Ok(Some(removed_inode)) => {
+                trace!(ino, "removing inode from superblock");
+                let parent_ino = removed_inode.parent();
+
+                if let Some((parent, _)) = inodes.get_mut(&parent_ino) {
+                    let mut parent_state = parent.get_mut_inode_state_no_check();
+                    let InodeKindData::Directory {
+                        children,
+                        writing_children,
+                        ..
+                    } = &mut parent_state.kind_data
+                    else {
+                        unreachable!("parent is always a directory");
+                    };
+                    if let Some(child) = children.get(removed_inode.name()) {
+                        // Don't accidentally remove a newer inode (e.g. remote shadowing local)
+                        if child.ino() == ino {
+                            children.remove(removed_inode.name());
+                        }
+                    }
+                    writing_children.remove(&ino);
                 } else {
                     // Should be impossible for this to fail (VFS inodes reference their parent, so
                     // children need to be freed first), but let's not crash in a `forget` function...
                     debug_assert!(false, "children should be forgotten before parents");
-                    return;
                 }
-            };
-            let mut parent_state = parent.get_mut_inode_state_no_check();
-            let InodeKindData::Directory {
-                children,
-                writing_children,
-                ..
-            } = &mut parent_state.kind_data
-            else {
-                unreachable!("parent is always a directory");
-            };
-            if let Some(child) = children.get(inode.name()) {
-                // Don't accidentally remove a newer inode (e.g. remote shadowing local)
-                if child.ino() == ino {
-                    children.remove(inode.name());
-                }
-            }
-            writing_children.remove(&ino);
 
-            if let Ok(state) = inode.get_inode_state() {
-                metrics::counter!("metadata_cache.inode_forgotten_before_expiry")
-                    .increment(state.stat.is_valid().into());
-            };
+                drop(inodes);
+
+                if let Ok(state) = removed_inode.get_inode_state() {
+                    metrics::counter!("metadata_cache.inode_forgotten_before_expiry")
+                        .increment(state.stat.is_valid().into());
+                };
+            }
+            Ok(None) => {}
+            Err(_) => {
+                // Inode was already removed --> error
```

Comment:
Should add this back?
```suggestion
                debug_assert!(
                    false,
                    "forget should not be called on inode already removed from superblock"
                );
```